### PR TITLE
Stop elemental-system-agent when the node is ready

### PR DESCRIPTION
**Pull request by davidcassany**
Fixes rancher/elemental#374

Signed-off-by: David Cassany <dcassany@suse.com>

### CODE REVIEW

File: pkg/controllers/machineinventoryselector/bootstrap.go [hunk 1]
```diff
@@ -85,22 +107,24 @@ func (h *handler) bootstrapReadyHandler(obj *v1beta1.MachineInventorySelector, s
...
+						"--no-block",
 						"stop",
 						"elemental-system-agent",
 					},
 				},
 			},
 		},
 	}
-	return json.Marshal(stopElementalAgent)
+	plan, _ := json.Marshal(stopElementalAgent)
```
Comment (Itxaka) [inline pkg/controllers/machineinventoryselector/bootstrap.go:126]: should check here for error? otherwise we may be returning empty plan

**Review by alexander-demicev — COMMENTED** [review]
This will have to be backported to the new logic based on kubebuilder, I created a special label for such PRs.

File: pkg/controllers/machineinventoryselector/bootstrap.go [hunk 2]
```diff
@@ -76,6 +77,27 @@ func (h *handler) bootstrapReadyHandler(obj *v1beta1.MachineInventorySelector, s
...
+		logrus.Info("bootstrap plan succeeded, setting plan to stop elemental-system-agent")
+		plan := getStopElementalAgentPlan()
+
+		planSecret, err := h.SecretCache.Get(inventory.Status.Plan.SecretRef.Namespace, inventory.Status.Plan.SecretRef.Name)
+		if err != nil {
+			logrus.Warn("failed to get plan secret, stop elemental-system-agent plan is not set")
+			return status, nil
+		}
+
+		planSecret.Data["plan"] = plan
```
Comment (Itxaka) [inline pkg/controllers/machineinventoryselector/bootstrap.go:95]: Im wondering, does elemental-system-agent and rancher-system-agent use the same secret for plan deploying? Could this override the plan set for the rancher-system-agent if they are using the same plan?

File: pkg/controllers/machineinventoryselector/bootstrap.go [hunk 3]
```diff
@@ -76,6 +77,27 @@ func (h *handler) bootstrapReadyHandler(obj *v1beta1.MachineInventorySelector, s
...
+
+		_, err = h.Secrets.Update(planSecret)
```
Comment (alexander-demicev) [inline pkg/controllers/machineinventoryselector/bootstrap.go:97]:
```suggestion
		if _, err := h.Secrets.Update(planSecret); if err != nil {
```

File: pkg/controllers/machineinventoryselector/bootstrap.go (hunk 2 above)
Comment (alexander-demicev) [reply pkg/controllers/machineinventoryselector/bootstrap.go:95]: machineinventory controller creates a secret, this secret is later picked up by machineinventoryselector controller and populated with a plan.

Comment (davidcassany) [reply pkg/controllers/machineinventoryselector/bootstrap.go:95]: > Im wondering, does elemental-system-agent and rancher-system-agent use the same secret for plan deploying? Could this override the plan set for the rancher-system-agent if they are using the same plan?

They are not using the same plan, that's why we need both. System agent listens only a single plan hence there is no easy way of making a handover from one to the other. We have no or little control over the rancher-system-agent plan and its setup. This is all managed by the bootstrap.sh script.

File: pkg/controllers/machineinventoryselector/bootstrap.go (hunk 1 above)
Comment (davidcassany) [reply pkg/controllers/machineinventoryselector/bootstrap.go:126]: Yes we can, however this is an static value. That's why I omitted it here.

**Review by alexander-demicev — APPROVED** [review]

**Pull request merged**